src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs

```diff
@@ -0,0 +1,184 @@
+                // look for a single remaining one
+                foreach (var member in members)
+                {
+                    if (member is MethodSymbol { IsOverride: false } method)
+                    {
+                        if (withMethod is null)
+                        {
+                            withMethod = method;
+                        }
+                        else
+                        {
+                            withMethod = null;
+                            break;
+                        }
+                    }
+                }
+            }
+
+            HashSet<DiagnosticInfo>? useSiteDiagnostics = null;
+            var lookupResult = LookupResult.GetInstance();
+            ImmutableArray<Symbol?> withMembers;
+
+            if (withMethod is null)
+            {
+                diagnostics.Add(ErrorCode.ERR_NoSingleWithMethod, syntax.Receiver.Location, receiverType);
+                withMembers = default;
+            }
+            else
+            {
+                // Check return type
+                if (!withMethod.ReturnType.IsEqualToOrDerivedFrom(
+                    receiverType,
+                    TypeCompareKind.ConsiderEverything,
+                    ref useSiteDiagnostics))
+                {
+                    diagnostics.Add(
+                        ErrorCode.ERR_ContainingTypeMustDeriveFromWithReturnType,
+                        syntax.Receiver.Location,
+                        receiverType,
+                        withMethod.ReturnType);
+                }
+
+                // Build WithMethod member list
+                var matchingMembers = ArrayBuilder<Symbol?>.GetInstance(withMethod.ParameterCount); 
+                foreach (var p in withMethod.Parameters)
+                {
+                    this.LookupMembersInType(
+                        lookupResult,
+                        receiverType,
+                        p.Name,
+                        arity: 0,
+                        basesBeingResolved: null,
+                        options: LookupOptions.Default,
+                        originalBinder: this,
+                        diagnose: false,
+                        useSiteDiagnostics: ref useSiteDiagnostics);
+
+                    Symbol? member;
+                    if (!lookupResult.IsSingleViable)
+                    {
+                        diagnostics.Add(
+                            ErrorCode.ERR_WithParameterWithoutMatchingMember,
+                            syntax.Receiver.Location,
+                            receiverType,
+                            p.Name);
+                        member = null;
+                    }
+                    else
+                    {
+                        member = lookupResult.SingleSymbolOrDefault;
+                        Debug.Assert(member.Kind == SymbolKind.Field || member.Kind == SymbolKind.Property);
```

Comment:
>Assert [](start = 30, length = 6)

I don't see a justification for this assertion.  What if it is an enumeration type?  What if it is an event?